docs/enhancements/cluster_pool_hot_spares.md

```diff
@@ -0,0 +1,111 @@
+# Maintaining Active Clusters in a ClusterPool
+
+[HIVE-1576](https://issues.redhat.com/browse/HIVE-1576)
+
+- [Maintaining Active Clusters in a ClusterPool](#maintaining-active-clusters-in-a-clusterpool)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+    - [User Story](#user-story)
+  - [Proposals](#proposals)
+    - [ClusterClaim Prioritizes Active Clusters](#clusterclaim-prioritizes-active-clusters)
+    - [ClusterClaim Must Account For `hibernateAfter`](#clusterclaim-must-account-for-hibernateafter)
+    - [Don't Hibernate New ClusterDeployments](#dont-hibernate-new-clusterdeployments)
+    - [`activeCount`/`activePercent`](#activecountactivepercent)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Test Plan](#test-plan)
+  - [Alternatives](#alternatives)
+
+## Summary
+
+Support configuring ClusterPools such that some number of clusters (0 <= N <= all) are Active for some amount of time (0 <= T <= eternity) before hibernating.
+
+## Motivation
+
+Today ClusterPools immediately hibernate new clusters.
+This means a ClusterClaim is guaranteed to suffer at least Resuming time before the cluster is available for use.
+While this is still faster than having to create a cluster from scratch, it is a potentially-unnecessary delay.
+As an example, CI runs could be reduced by ~5m if a ClusterClaim were able to grab an Active cluster rather than having to Resume a Hibernating one.
+
+### Goals
+
+Ability to tune a ClusterPool such that a ClusterClaim gets an Active cluster immediately in the common case.
+
+### Non-Goals
+
+Solving world peace.
+
+
+### User Story
+
+As a consumer of a (properly tuned) ClusterPool,
+
+I want ClusterClaim to give me an active cluster immediately in the common case,
+
+So that I can get my work done more quickly.
+
+(Ideally without sacrificing the cost savings of having some clusters in the pool hibernating.)
+
+## Proposals
+
+Several implementation alternatives have already come to mind.
+These are not necessarily mutually exclusive.
+
+### ClusterClaim Prioritizes Active Clusters
+
+For most/all of the options below, it makes sense to add logic to ClusterClaim to preferentially grab an Active cluster first.
+The simplest implementation is to sort the list of ClusterDeployments into three buckets: Active, Resuming, and Hiberating, claimed in that order.
+A smarter implementation could take into account when Resuming clusters were asked to resume, with the most recent at the end, on the theory that the least-recently-resumed will be ready first.
+
+### ClusterClaim Must Account For `hibernateAfter`
+
+This applies to most/all of the options below.
+
+If the ClusterPool has `hibernateAfter` set, that clock starts running as soon as the ClusterDeployment is ready.
+(TODO: Confirm this -- does the clock start at creation, or `installed == True`, or something else?)
```

Comment:
Today, the hibernation_controller adds `hibernateAfter` to a base time to decide whether to hibernate.
To determine the base time, it looks for the `Hibernating` Condition on the CD.
- If `Unknown`, uses the `status.installedTimestamp` (when installation completed).
- If `False`, uses the `lastTransitionTime` of the Condition (when it woke up from the most recent previous hibernation).

What we could do here is add another Condition, e.g. `Claimed`, indicating claim assignment status. (We already have the equivalent on the claim.) Then change the above to:
- If the CD is part of a ClusterPool
  - If it is unclaimed, ignore `hibernateAfter` [1] -- never hibernate.
  - If it is claimed, use as the base time the _later_ of
    - the `lastTransitionTime` of the new `Claimed` Condition
    - the above algorithm's base time (this accounts for hibernating/resuming during the normal post-claim lifecycle of the CD).
- Else the above algorithm as is

[1] This also means a manually-resumed unclaimed CD from a pool will never hibernate at the behest of `hibernateAfter`. IMO we can live with that. If you're using `hibernateAfter` with ClusterPools, you really oughtta let the code manage things.

Reply:
Added to doc.